core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -40,22 +40,30 @@ import java.util.concurrent.locks.ReentrantReadWriteLock
 private[server] class MetadataCache(brokerId: Int) extends Logging {
   private val stateChangeLogger = KafkaController.stateChangeLogger
   private val cache = mutable.Map[String, mutable.Map[Int, PartitionStateInfo]]()
-  private var aliveBrokers = Map[Int, Broker]()
+  private val aliveBrokers = mutable.Map[Int, Broker]()
+  private val aliveNodes = mutable.Map[Int, collection.Map[SecurityProtocol, Node]]()
   private val partitionMetadataLock = new ReentrantReadWriteLock()
 
   this.logIdent = "[Kafka Metadata Cache on broker %d] ".format(brokerId)
 
   private def getAliveEndpoints(brokers: Iterable[Int], protocol: SecurityProtocol): Seq[Node] = {
-    brokers.toSeq.flatMap(aliveBrokers.get(_).map(_.getNode(protocol)))
+    val result = new mutable.ArrayBuffer[Node](math.min(aliveBrokers.size, brokers.size))
+    brokers.foreach { brokerId =>
+      getAliveEndpoint(brokerId, protocol).foreach(result +=)
```

Comment:
Yeah, it is known that for performance sensitive parts of the code, imperative code with mutable collections performs better.